File: web/src/lib/instance.ts
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (C) 2026 Serraniel and the Sendan contributors

/**
 * What an instance says it permits.
 *
 * Everything an instance would refuse was previously discoverable only by
 * trying: the size it accepts, how long an upload may live, whether it may live
 * forever, whether a limit is compulsory. Somebody deciding whether to use an
 * instance at all - a stranger's instance in particular - should be able to see
 * the rules rather than discover them by being refused.
 *
 * ## What this is worth
 *
 * An instance can state whatever it likes here, and nothing binds it to the
 * truth. This is a convenience rather than a guarantee, and the interface says
 * so where it shows it. The properties that do not depend on the instance being
 * honest are the cryptographic ones: the key is derived and used in this
 * browser whatever this endpoint claims.
 *
 * ## Absent values
 *
 * Every field can be absent, and absent means "not said" rather than a default
 * invented here. An interface that filled in a plausible number would be making
 * exactly the claim this module exists to avoid.
 */

import type { Transport } from "$lib/tus";

/** The policy an instance reports. Any field may be absent. */
export interface InstancePolicy {
  maxUploadSize: number | null;
  defaultTtlSeconds: number | null;
  maxTtlSeconds: number | null;
  allowInfiniteTtl: boolean | null;
  requireLimit: boolean | null;
  defaultMaxDownloads: number | null;
  compatEnabled: boolean | null;
}

/** What is known when the instance has said nothing. */
export const nothingKnown: InstancePolicy = {
  maxUploadSize: null,
  defaultTtlSeconds: null,
  maxTtlSeconds: null,
  allowInfiniteTtl: null,
  requireLimit: null,
  defaultMaxDownloads: null,
  compatEnabled: null,
};

/**
 * Asks the instance for its policy.
 *
 * Never throws. An instance that cannot be reached, or that answers with
 * something else entirely, leaves the interface exactly as it was: a page that
 * refused to work because it could not read a policy would be worse than one
 * that does not know it.
 */
export async function fetchInstancePolicy(
  transport: Transport = {},
  origin = "",
): Promise<InstancePolicy> {
  const doFetch = transport.fetch ?? fetch;

  try {
    const response = await doFetch(`${origin}/api/instance`, {
      ...(transport.signal ? { signal: transport.signal } : {}),
    });
    if (!response.ok) return nothingKnown;
    return parsePolicy(await response.json());
  } catch {
    return nothingKnown;
  }
}

/**
 * Reads a policy out of whatever arrived.
 *
 * Field by field, because this is a value from the network and an instance is
 * free to send anything. A field of the wrong type is treated as unsaid rather
 * than coerced: "3 days" coerced to a number is not a mistake worth passing on
 * to somebody choosing a retention period.
 */
export function parsePolicy(value: unknown): InstancePolicy {
  if (typeof value !== "object" || value === null) return nothingKnown;
  const v = value as Record<string, unknown>;

  return {
    maxUploadSize: positive(v.maxUploadSize),
    defaultTtlSeconds: positive(v.defaultTtlSeconds),
    maxTtlSeconds: positive(v.maxTtlSeconds),
    allowInfiniteTtl: boolean(v.allowInfiniteTtl),
    requireLimit: boolean(v.requireLimit),
    // Zero is meaningful here - it is how "no limit" is expressed - so this one
    // admits it while the durations above do not.
    defaultMaxDownloads: whole(v.defaultMaxDownloads),
    compatEnabled: boolean(v.compatEnabled),
  };
}

function positive(value: unknown): number | null {
  return typeof value === "number" && Number.isFinite(value) && value > 0 ? value : null;
}

function whole(value: unknown): number | null {
  return typeof value === "number" && Number.isInteger(value) && value >= 0 ? value : null;
}

function boolean(value: unknown): boolean | null {
  return typeof value === "boolean" ? value : null;
}

/** A duration somebody can read, from seconds. */
export function formatDuration(seconds: number): string {
  const units: Array<[number, string]> = [
    [86400, "day"],
    [3600, "hour"],
    [60, "minute"],
  ];

  for (const [size, name] of units) {
    if (seconds >= size) {
      const n = Math.round(seconds / size);
      return `${n} ${name}${n === 1 ? "" : "s"}`;
    }
  }
  return `${seconds} second${seconds === 1 ? "" : "s"}`;
}

File: internal/httpapi/instance.go
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (C) 2026 Serraniel and the Sendan contributors

package httpapi

import (
	"encoding/json"
	"net/http"
	"time"
)

// Instance is what an instance says about the rules it enforces.
//
// Everything here is policy: what an upload may ask for and what it will be
// refused. None of it describes the deployment, and that boundary is the
// interesting part of this type rather than an implementation detail.
//
// # What is deliberately absent
//
// Which storage backend is in use, which database, whether at-rest keys are
// wrapped under a master key, how many proxies stand in front, and what the
// rate limits are. None of it changes what a person may upload, and all of it
// helps somebody attacking the instance more than it helps somebody using it.
//
// # What this is worth
//
// An instance can state whatever it likes here; nothing binds it to the truth.
// It is a convenience, not a guarantee, and the interface says so. The
// properties that do not depend on the instance being honest are the
// cryptographic ones - the key is derived and used in the browser, whatever
// this endpoint claims.
type Instance struct {
	// MaxUploadSize is the largest single upload, in bytes. Zero means the
	// instance sets no limit of its own.
	//
	// Also advertised as Tus-Max-Size on the upload endpoint, which is where
	// the client reads it during a transfer. Repeated here so that everything
	// a person is shown before choosing a file comes from one answer.
	MaxUploadSize int64 `json:"maxUploadSize"`

	// DefaultTTLSeconds is how long an upload lives when the uploader does not
	// choose, and MaxTTLSeconds is the longest they may choose.
	DefaultTTLSeconds int64 `json:"defaultTtlSeconds"`
	MaxTTLSeconds     int64 `json:"maxTtlSeconds"`

	// AllowInfiniteTTL is whether an upload may be asked to never expire.
	AllowInfiniteTTL bool `json:"allowInfiniteTtl"`

	// RequireLimit is whether every upload must carry either a deadline or a
	// download limit. It can only bind where unlimited retention is allowed,
	// since otherwise every upload already has a deadline.
	RequireLimit bool `json:"requireLimit"`

	// DefaultMaxDownloads is the download limit applied when the uploader does
	// not choose one. Zero means no limit.
	DefaultMaxDownloads int `json:"defaultMaxDownloads"`

	// CompatEnabled reports whether the third-party compatibility endpoints are
	// served.
	//
	// Published because it is the one setting that changes what protection an
	// upload can have: through those endpoints the instance checks the password
	// rather than the password contributing to the key. Somebody choosing an
	// instance is entitled to know that before they use it, not after.
	CompatEnabled bool `json:"compatEnabled"`
}

// handleInstance reports the policy an instance enforces.
func handleInstance(in Instance) http.HandlerFunc {
	return func(w http.ResponseWriter, _ *http.Request) {
		w.Header().Set("Content-Type", "application/json; charset=utf-8")
		// Short, not never: the answer changes only when an operator changes
		// configuration and restarts, but a page that caches it for a day would
		// show a limit that no longer applies.
		w.Header().Set("Cache-Control", "public, max-age=60")

		if err := json.NewEncoder(w).Encode(in); err != nil {
			// The header is already written, so there is nothing useful to say
			// to the client. Nothing here is secret and nothing is identifying,
			// so there is nothing to redact either.
			return
		}
	}
}

// describeInstance builds the report from the options an instance was given.
func describeInstance(opts Options) Instance {
	in := Instance{
		MaxUploadSize: opts.MaxUploadSize,
		CompatEnabled: opts.Compat != nil,
	}

	// A service is absent in a backend-only assembly and in tests that exercise
	// the HTTP surface alone. Reporting zeroes there would be describing a
	// policy nobody set, so the retention fields stay at their zero values and
	// mean the same thing they mean everywhere else: no answer.
	if opts.Uploads == nil {
		return in
	}

	policy := opts.Uploads.Policy()
	in.DefaultTTLSeconds = seconds(policy.DefaultTTL)
	in.MaxTTLSeconds = seconds(policy.MaxTTL)
	in.AllowInfiniteTTL = policy.AllowInfiniteTTL
	in.RequireLimit = policy.RequireLimit
	in.DefaultMaxDownloads = policy.DefaultMaxDownloads
	return in
}

// seconds converts a duration for the wire, rounding down.
//
// Rounding down rather than to nearest: a maximum reported as longer than it is
// produces a request the instance then refuses, which is the failure this
// endpoint exists to prevent.
func seconds(d time.Duration) int64 {
	if d <= 0 {
		return 0
	}
	return int64(d / time.Second)
}
